qdev_wrappers/fitting/Fitter.py

```diff
@@ -0,0 +1,289 @@
+            p0 = getattr(fitclass, 'guess')(**data_dict)
+            return p0
+        elif p0 != None:
+            p0 = p0
+            return p0
+        else:
+            return "Could not find guess parameters for fit."
+
+    def find_parameter_units(self, fitclass, unit_dict):
+
+        unit_templates = fitclass.p_units
+        param_units = []
+
+        for template in unit_templates:
+            template = list(template)
+            for i in range(len(template)):
+                for variable in unit_dict.keys():
+                    if template[i] == variable:
+                        template[i] = unit_dict[variable]
+            unit = "".join(template)
+            param_units.append(unit)
+
+        return param_units
+
+    def perform_fit(self, data_dict, data, fitclass, setpoints, p0, param_units, **kwargs):
+
+        fit = {}
+        fit['parameters'] = {}
+        fit['start_params'] = {}
+
+        xdata = data_dict[self.input_vars[0]]
+        ydata = data_dict[self.output_var[0]]
+
+        p_guess = self.find_start_parameters(fitclass, p0, data_dict)
+        popt, pcov = curve_fit(fitclass.fun, xdata, ydata, p0=p_guess, **kwargs)
+
+        for parameter in fitclass.p_labels:
+            fit['parameters'][parameter] = {'value': popt[fitclass.p_labels.index(parameter)]}
+            fit['parameters'][parameter]['cov'] = pcov[fitclass.p_labels.index(parameter)]
+            fit['parameters'][parameter]['unit'] = param_units[fitclass.p_labels.index(parameter)]
+            fit['start_params'][parameter] = p_guess[fitclass.p_labels.index(parameter)]
+
+        # make an dictionary containing arrays of each parameter value with same dimension as data (for estimate)
+        all_params = {}
+        for parameter in fitclass.p_labels:
+            all_params[parameter] = []
+            for datapoint in xdata:
+                all_params[parameter].append(fit['parameters'][parameter]['value'])
+            all_params[parameter] = np.array(all_params[parameter])
+        self.full_parameter_dict = all_params
+
+        return fit
+
+    def estimate_function_values(self, fitclass, data, data_dict):
+
+        estimate = {}
+
+        input_data = data_dict
+        for variable in self.output_var:
+                del input_data[variable]
+
+        input_parameters = self.full_parameter_dict
+
+        # Calculate estimated values for output based on function and store in estimate['values']
+        estimate['values'] = fitclass.fun(**input_data, **input_parameters)
+
+        # Save estimate
+        estimate['name'] = '{}_estimate'.format(self.output_dataname[0])
+        estimate['label'] = '{} estimate'.format(data[self.output_dataname[0]]['label'])
+        estimate['unit'] = data[self.output_dataname[0]]['unit']
+        estimate['parameters'] = self.full_parameter_dict   #not sure this is used for anything
+
+        return estimate
+
+
+class Fitter_2Ddata_1Dfunction(Fitter1D):
+
```

Comment:
again a comment describing the class would be great, I also wonder if a class is the best way to implement this. I feel like we want to make something general enough that we can say 'fit to these n parameters for each of these m setpoint parameters and give me a dataset with m values for each output param of the fit'. One instance of this is the 2D dataset and a 1D fit but as soon as we have more setpoint parameters in a dataset or are fitting to more than 2 parameters the number of these classes we need will blow up, maybe we could just make one general one and push the specific stuff into the plotting which is anyways more of a 'convenience wrapper'

Reply:
I'm not sure how to go about making them all one class, but I agree that that would be optimal. It seems like different reorganization of the data is needed depending on whether there are setpoints, and how many variables act as set points.